src/lib/gadgets/foreign-field.ts

```diff
@@ -431,3 +456,216 @@ function combine2([x0, x1]: bigint3 | [bigint, bigint]) {
+    this.#summands.push(y);
+    return this;
+  }
+
+  sub(y: Field3) {
+    assert(this.#result === undefined, 'sum already finished');
+    this.#ops.push(-1n);
+    this.#summands.push(y);
+    return this;
+  }
+
+  #return(x: Field3) {
+    this.#result = x;
+    return x;
+  }
+
+  isConstant() {
+    return this.#summands.every(Field3.isConstant);
+  }
+
+  finish(f: bigint, isChained = false) {
+    assert(this.#result === undefined, 'sum already finished');
+    let signs = this.#ops;
+    let n = signs.length;
+    if (n === 0) return this.#return(this.#summands[0]);
+
+    // constant case
+    if (this.isConstant()) {
+      return this.#return(sum(this.#summands, signs, f));
+    }
+
+    // provable case
+    let x = this.#summands.map(toVars);
+    let result = x[0];
+
+    for (let i = 0; i < n; i++) {
+      ({ result } = singleAdd(result, x[i + 1], signs[i], f));
+    }
+    if (!isChained) Gates.zero(...result);
+
+    this.#result = result;
+    return result;
+  }
+
+  // TODO this is complex and should be removed once we fix the ffadd gate to constrain all limbs individually
+  finishForMulInput(f: bigint, isChained = false) {
+    assert(this.#result === undefined, 'sum already finished');
+    let signs = this.#ops;
+    let n = signs.length;
+    if (n === 0) return this.#return(this.#summands[0]);
+
+    // constant case
+    if (this.isConstant()) {
+      return this.#return(sum(this.#summands, signs, f));
+    }
+
+    // provable case
+    let xs = this.#summands.map(toVars);
+
+    // since the sum becomes a multiplication input, we need to constrain all limbs _individually_.
+    // sadly, ffadd only constrains the low and middle limb together.
+    // we could fix it with a RC just for the lower two limbs
+    // but it's cheaper to add generic gates which handle the lowest limb separately, and avoids the unfilled MRC slot
+    let f0 = f & lMask;
+
+    // generic gates for low limbs
+    let x0 = xs[0][0];
+    let x0s: Field[] = [];
+    let overflows: Field[] = [];
+    let xRef = Unconstrained.witness(() => Field3.toBigint(xs[0]));
+
+    for (let i = 0; i < n; i++) {
+      // compute carry and overflow
+      let [carry, overflow] = exists(2, () => {
+        // this duplicates some of the logic in singleAdd
+        let x = xRef.get();
+        let x0 = x & lMask;
+        let xi = toBigint3(xs[i + 1]);
+        let sign = signs[i];
+
+        // figure out if there's overflow
+        x += sign * combine(xi);
+        let overflow = 0n;
+        if (sign === 1n && x >= f) overflow = 1n;
+        if (sign === -1n && x < 0n) overflow = -1n;
+        if (f === 0n) overflow = 0n;
+        xRef.set(x - overflow * f);
+
+        // add with carry, only on the lowest limb
+        x0 = x0 + sign * xi[0] - overflow * f0;
+        let carry = x0 >> l;
+        return [carry, overflow];
+      });
+      overflows.push(overflow);
+
+      // constrain carry
+      assertOneOf(carry, [0n, 1n, -1n]);
```

Comment:
don't you need to constrain `overflow` also?

Reply:
no, see https://github.com/o1-labs/o1js/pull/1262#discussion_r1415877018